utils-zookeeper/src/main/java/cz/o2/proxima/utils/zookeeper/ZKGlobalWatermarkTracker.java

```diff
@@ -0,0 +1,458 @@
+    getNodeVersionToFuture(path, res);
+    return res;
+  }
+
+  private void getNodeVersionToFuture(String path, CompletableFuture<Integer> res) {
+    client()
+        .getData(
+            path,
+            false,
+            (code, p, ctx, bytes, stat) -> {
+              if (code == Code.CONNECTIONLOSS.intValue()
+                  || code == Code.SESSIONEXPIRED.intValue()) {
+                handleConnectionLoss(() -> getNodeVersionToFuture(path, res));
+              } else if (code != Code.OK.intValue()) {
+                res.completeExceptionally(
+                    new RuntimeException(
+                        String.format("Error fetching version of %s: %d", path, code)));
+              } else {
+                res.complete(stat.getVersion());
+              }
+            },
+            null);
+  }
+
+  private void handleConnectionLoss(Runnable retry) {
+    client = null;
+    retry.run();
+  }
+
+  private void handleBadVersion(String path, CompletableFuture<Void> res, Runnable onSuccess) {
+    getNodeVersion(path)
+        .whenComplete(
+            (v, exc) -> {
+              if (exc != null) {
+                res.completeExceptionally(exc);
+              } else {
+                updateVersion = v;
+                onSuccess.run();
+              }
+            });
+  }
+
+  private synchronized void createParentIfNotExists(String node) throws InterruptedException {
+    try {
+      if (!parentCreated) {
+        client().create(node, new byte[] {}, Ids.OPEN_ACL_UNSAFE, CreateMode.CONTAINER);
+        createWatchForParentNode(node);
+        parentCreated = true;
+      }
+    } catch (KeeperException ex) {
+      if (ex.code() == Code.CONNECTIONLOSS || ex.code() == Code.SESSIONEXPIRED) {
+        client = null;
+        createParentIfNotExists(node);
+      } else if (ex.code() != Code.NODEEXISTS) {
+        throw new RuntimeException(ex);
+      } else {
+        createWatchForParentNode(node);
+      }
+    }
+  }
+
+  private static byte[] longAsBytes(long watermark) {
+    return String.valueOf(watermark).getBytes(StandardCharsets.UTF_8);
+  }
+
+  private static long longFromBytes(byte[] bytes) {
+    return Long.parseLong(new String(bytes, StandardCharsets.UTF_8));
+  }
+
+  private String getParentNode() {
+    return parentNode + this.trackerName;
+  }
+
+  private synchronized void updatePartialWatermark(String name, long watermark) {
+    partialWatermarks.compute(
+        name, (k, value) -> Math.max(value == null ? Long.MIN_VALUE : value, watermark));
+
+    globalWatermark.set(
+        partialWatermarks.values().stream().min(Long::compare).orElse(Long.MIN_VALUE));
+
+    Optional.ofNullable(updatedFutures.remove(name))
+        .ifPresent(l -> l.forEach(f -> f.complete(null)));
+  }
+
+  private ZooKeeper client() {
+    if (client == null) {
+      synchronized (this) {
+        if (client == null) {
+          client = createNewZooKeeper();
```

Comment:
maybe we can cache and reuse client instance in case of multiple instances in as single JVM?

Reply:
We could, but I strongly prefer doing such optimizations when we see that it is actually a problem. It certainly brings problems which might not be worth it.